### PR TITLE
Added MyPy Support

### PR DESCRIPTION
**Pull request by GameDungeon**
## Overview: What does this pull request change?

I added mypy support, including the pre-commit hook, and also fixed some issues mypy brought up.

Many things are ignored, it is a task for future pulls to fix the issues.

## Reviewer Checklist
- [x] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [x] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [x] If applicable: newly added functions and classes are tested

### CODE REVIEW

**Comment by GameDungeon** [comment]
While we have been using mypy to fix issues in manim for a while, I think that it's important to be officially part of manim's developer tool kit.

File: manim/_config/__init__.pyi [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+from .utils import ManimConfig
+
+logger: Any
+console: Console
+error_console: Console
+config: Any
+frame: Any
+
+def tempconfig(temp: Union[ManimConfig, dict]) -> _GeneratorContextManager: ...
```
Comment (behackl) [inline manim/_config/__init__.pyi:14]: Why can these types not be added to the actual code file?

**Review by behackl — COMMENTED** [review]
I have no experience with MyPy at all, it would be great if someone else could look at the concrete changes here.

I don't particularly like the addition of 7 new dependencies.

File: manim/_config/__init__.pyi (hunk 1 above)
Comment (GameDungeon) [reply manim/_config/__init__.pyi:14]: The capitalized colors are generated in runtime so it does not work.

**Comment by GameDungeon** [comment]
The new dependencies are Dev Dependecys, and you can't get much more maintained than under the python github account + microsoft

File: manim/_config/__init__.pyi (hunk 1 above)
Comment (behackl) [reply manim/_config/__init__.pyi:14]: This comment is not about `color.pyi`

Comment (GameDungeon) [reply manim/_config/__init__.pyi:14]: Ohh sorry. Mypy was interacting weirdly and this seemed to be the most direct solution.

**Comment by hydrobeam** [comment]
What's the point of mypy? I'm not even sure why we need it.

**Comment by GameDungeon** [comment]
Static Type checking. It finds bugs in our documentation & code. If we mark functions for the wrong types it can mislead people. There have already been several prs fixing mypy errors.

**Comment by christopher-besch** [comment]
This looks like something we could really use.
Adding the type-hint dependencies shouldn't be an issue; they aren't that big and are required anyway if we want to make Manim type-safe. I don't think there is any way around MyPy once we move to the new dev-branch.

What I don't understand is why the `.pyi` files were changed.

**Comment by GameDungeon** [comment]
This pull is old enough that if we plan to merge it, it might be better to delete the pyi files.

File: .mypy.ini [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ignore_errors = True
+
+[mypy-manim._config.*]
+ignore_errors = True
+
+[mypy-manim.utils.*]
+ignore_errors = True
+
+[mypy-manim.animation.*]
+ignore_errors = True
```
Comment (christopher-besch) [inline .mypy.ini:32]:
```suggestion
[mypy-manim.__main__,manim.renderer.*,manim.camera.*,manim.scene.*,manim.cli.cfg.*,manim.mobject.*,manim._config.*,manim.utils.*,manim.animation.*]
ignore_errors = True
```

Why not do it like this?

**Comment by christopher-besch** [comment]
I gave it a run and this was the result:
image

Did I do something wrong?

**Comment by GameDungeon** [comment]
Try only running it on manim

**Comment by christopher-besch** [comment]
> Try only running it on manim's folder.

I still get this error:
image

**Comment by christopher-besch** [comment]
We should also add this to the `.pre-commit-config.yaml`.

**Comment by GameDungeon** [comment]
I've been away for a while, sorry for the really late response. I believe it's too early to add it to pre-commit.That one error was introduced since I made this.

File: .mypy.ini [hunk 3]
```diff
@@ -0,0 +1,95 @@
+[mypy]
+python_version = 3.7
+show_error_codes = True
+
+# ---------------- Needs better typing, project for a future pull -----
+
```
Comment (christopher-besch) [inline .mypy.ini:6]:
```suggestion
# ignore most files; should be checked once proper types have been implemented
```

File: .mypy.ini [hunk 4]
```diff
@@ -0,0 +1,95 @@
+[mypy]
+python_version = 3.7
```
Comment (christopher-besch) [inline .mypy.ini:2]: Can't we use a newer version?

**Review by christopher-besch — CHANGES_REQUESTED** [review]

**Comment by christopher-besch** [comment]
Could you also merge any new changes to Manim and check if MyPy runs?

File: .mypy.ini (hunk 4 above)
Comment (GameDungeon) [reply .mypy.ini:2]: That's the current lowest version of python that manim actively supports last time I checked.

Comment (christopher-besch) [reply .mypy.ini:2]: So we should use the lowest possible version? Why so, if I may ask?

Comment (GameDungeon) [reply .mypy.ini:2]: I'm not 100% sure what that setting does, but possibly breaking support for versions we still claim we support tends to be a bad thing.

**Comment by GameDungeon** [comment]
I think everything is up to data and functioning now :) I fixed the logger error as well.

File: .mypy.ini (hunk 4 above)
Comment (GameDungeon) [reply .mypy.ini]: Ah, I don't remember why it was added, but it's optional and does not really affect the end result so I removed it.

**Comment by GameDungeon** [comment]
I also managed to get the mypy pre-commit hook working :)

**Comment by christopher-besch** [comment]
I ran it on my machine and everything seems to be working. With pre-commit it is also being run in CI.
I think we can merge this.

Thanks for the contribution!

**Review by christopher-besch — APPROVED** [review]

**Review by Darylgolden — APPROVED** [review]
I think this looks good, thank you!

**Pull request merged**